processor.go

```diff
@@ -96,204 +116,193 @@ func (p *Processor[AC, OC, JC]) init() {
 	}
 	// Make a map of triggers to states so we can easily reference it
 	p.stateMap = map[string]State[AC, OC, JC]{}
-	for _, s := range p.states {
-		p.stateMap[s.TriggerState] = s
-	}
-	// get a list of return state names for use
-	p.stateNames = make([]string, 0, len(p.stateMap))
-	for k := range p.stateMap {
-		p.stateNames = append(p.stateNames, k)
-	}
-
-	// For each state, we need a channel of jobs
+	p.stateStatusMap = map[string]*StatusCount{}
+	p.stateWaitingJobsMap = map[string][]Job[JC]{}
 	p.stateChan = map[string]chan Job[JC]{}
+	p.sortedStateNames = []string{}
 
-	// Create the state chans
-	totalConcurrency := 0
 	for _, s := range p.states {
-		if s.Terminal {
-			continue
+		stateName := s.TriggerState
+
+		p.sortedStateNames = append(p.sortedStateNames, stateName)
+		p.stateMap[stateName] = s
+		p.stateStatusMap[stateName] = &StatusCount{
+			State:    stateName,
+			Terminal: s.Terminal,
 		}
-		p.stateChan[s.TriggerState] = make(chan Job[JC], s.Concurrency) // make a chan
-		totalConcurrency += s.Concurrency
+		// This is by-design unbuffered
+		p.stateChan[stateName] = make(chan Job[JC])
```

Comment:
So doing this will show any issues in the code that attempts to send to a channel awaiting jobs... is I think the point you were making?

Reply:
There are three positions a job can be in: 
1. Completed (Terminal)
2. Waiting to run
3. Running

We have one goroutine that is responsible for moving jobs between these states (think javascript event loop), and then a number of goroutines that are responsible for actual job running.

The key point is that the channels we use to pass jobs between the main "job manager" Goroutine and the runners should't need to contain any sort of buffering logic. We should only pass a job to a runner when we know it can run it. If it can't be run, the job should be buffered somewhere (right now, that is in the processor, but could very easily be in the StateExec itself), and when a "slot" frees up, we send the corresponding runner its next job.